### ISSUE TITLE
Adding Backward Control Flow Dependencies/Triggers

### ISSUE THREAD

**Issue by demmerichs** [pytorch/pytorch#74742]
### 🚀 The feature, motivation and pitch

I would like to be able to modify the DAG compute graph to trigger backward computation of certain nodes during a global backward pass, depending on other backward grad_fn calls of other nodes.

An example I would like to solve:
```pseudo
def module(data, scores):
    idxs = argsort(data)
    return data[idxs]

data, scores = model(input)
output = module(data, scores)
loss = criterion(output)
loss.backward()
```
Now during backward pass the grad_fn on the data input is called, but not on the scores tensor, as we used a non-differentiable process inside the module. However, I want to construct a custom gradient using `scores.register_hook(some_custom_grad_construct)` but the backward pass in this example will never call the grad_fn for scores, as it is not connected to the loss through the taped DAG.

### Alternatives

I could not find any documentated feature, that allows me to do this modification. I am currently thinking of implementing an `autograd.Function` as work-around which simply collects as input the tensors that should be dependent during backward, but passes during forward only the tensors without modification through, but I am afraid it might produce quite an overhead.

```diff
+class ControlGroup(autograd.Function):
+   def forward(*inputs):
+      return inputs
+   def backward(*grads):
+      return grads

def module(data, scores):
+   scores.register_hook(some_backward_hook)
+   data, scores = ControlGroup(data, scores)
    idxs = argsort(data)
    return data[idxs]

data, scores = model(input)
output = module(data, scores)
loss = criterion(output)
loss.backward()
```

I have not tested this yet, I would be happy to hear about thoughts on my workaround, possible other solutions that could currently be employed, and if an official support for such an feature is coming.

Thanks for your help.

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

**Comment by albanD** [comment]
Hi,

I think the custom Function is the right way to go here:
```python
class MyFn(autograd.Function):
    @staticmethod
    def forward(ctx, data, scores):
        # Save anything you might need later
        ctx.saved_for_backward(data, scores)
        # Just do the same as before
        idxs = argsort(data)
        return data[idxs]

    @staticmethod
    def backward(ctx, grad_out):
        data, scores = ctx.saved_tensors
        # The "real" gradient for the data
        grad_data = torch.zeros_like(data).scatter(idx, grad_out)
        # Some gradient for the scores
        grad_scores = torch.rand_like(scores)
        return grad_data, grad_scores

def module(data, scores):
    return MyFn.apply(data, scores)
```

I didn't run the code so you might need to clean up some typos there.
For more details on how to write custom Functions see https://pytorch.org/docs/stable/notes/extending.html
